### ISSUE TITLE
Update Flux in 1.x InfluxDB

### ISSUE THREAD

**Issue by nathanielc** [influxdata/influxdb#12872]
The InfluxDB 1.x repo depends on the 2.x repo in the form of the platform repo.

So in order to update Flux in InfluxDB 1.x we need to be able  to import InfluxDB 2.x as a dependency of 1.x.

This likely means that InfluxDB 1.x branch needs to understand Go modules.

RACI:
R: @jsternberg 	
A: @nathanielc 	
C: eng-leads
I: contributors and eng.

**Comment by jsternberg** [comment]
I'm going to consult with @mark-rushakoff about this before doing anything or consulting with a larger audience.

There's two options I see as happening.

1. Modify the v2 influxdb repository (on master) to use the `/v2` suffix for its module path. I think this is the best option, but it will be potentially disruptive at the moment. I need to also test if it works without actually tagging a version like the `v0.0.0` tags do.
2. Switch InfluxDB 1.x to use modules and abuse replace to rename the v2 version of influxdb to platform. I'm not sure this physically works because it will likely get confused with the v1 version.

I think 1 is the easiest and the one intended by the go modules developers and it should (hopefully) be pretty straightforward. It's likely also the only way this actually works.

**Comment by jsternberg** [comment]
I've proof of concepted using v2 for version 2 by creating a repository where a v1 lives and uses a v2 that's in the master branch with no tags so that will work perfectly even if it would be a bit disruptive.

**Comment by jsternberg** [comment]
Additional note: Vendor works perfectly when doing the above so that workflow will still work.

**Comment by jsternberg** [comment]
I talked with @mark-rushakoff about this and we believed it was likely the easiest method. Some additional questions that were brought up was whether we could get away with copying over the code so we had no dependencies on the v2 version in v1 and I think it'll likely be too complicated, but it is a possible thing we can do.

Here is a list of the references to platform:

```
Gopkg.lock:471:  name = "github.com/influxdata/platform"
Gopkg.lock:1114:    "github.com/influxdata/platform/models",
Gopkg.lock:1115:    "github.com/influxdata/platform/pkg/data/gen",
Gopkg.lock:1116:    "github.com/influxdata/platform/query/functions/inputs/storage",
Gopkg.lock:1117:    "github.com/influxdata/platform/storage/reads",
Gopkg.lock:1118:    "github.com/influxdata/platform/storage/reads/datatypes",
Gopkg.lock:1119:    "github.com/influxdata/platform/tsdb/cursors",
Gopkg.toml:1:required = ["github.com/influxdata/platform/pkg/data/gen"]
cmd/influxd/run/server.go:36:   "github.com/influxdata/platform/storage/reads"
cmd/store/query/query.go:20:    "github.com/influxdata/platform/storage/reads/datatypes"
flux/control/controller.go:8:   fstorage "github.com/influxdata/platform/query/functions/inputs/storage"
flux/functions/inputs/from.go:11:       "github.com/influxdata/platform/query/functions/inputs/storage"
internal/storage_store.go:8:    "github.com/influxdata/platform/storage/reads"
internal/storage_store.go:9:    "github.com/influxdata/platform/storage/reads/datatypes"
models/points.go:19:    "github.com/influxdata/platform/models"
prometheus/converters.go:13:    "github.com/influxdata/platform/storage/reads/datatypes"
services/httpd/handler.go:42:   "github.com/influxdata/platform/storage/reads"
services/httpd/handler.go:43:   "github.com/influxdata/platform/storage/reads/datatypes"
services/httpd/handler_test.go:39:      "github.com/influxdata/platform/storage/reads"
services/httpd/handler_test.go:40:      "github.com/influxdata/platform/storage/reads/datatypes"
services/storage/predicate_test.go:9:   "github.com/influxdata/platform/storage/reads"
services/storage/predicate_test.go:10:  "github.com/influxdata/platform/storage/reads/datatypes"
services/storage/series_cursor.go:12:   "github.com/influxdata/platform/storage/reads"
services/storage/series_cursor.go:13:   "github.com/influxdata/platform/storage/reads/datatypes"
services/storage/series_cursor_test.go:8:       "github.com/influxdata/platform/models"
services/storage/store.go:15:   "github.com/influxdata/platform/query/functions/inputs/storage"
services/storage/store.go:16:   "github.com/influxdata/platform/storage/reads"
services/storage/store.go:17:   "github.com/influxdata/platform/storage/reads/datatypes"
tsdb/cursor.go:7:       "github.com/influxdata/platform/tsdb/cursors"
```

Here is the list of dependencies referencing platform (that we would need to copy):

```
github.com/influxdata/platform/pkg/escape
github.com/influxdata/platform/models
github.com/influxdata/platform/telegraf/plugins
github.com/influxdata/platform/telegraf/plugins/inputs
github.com/influxdata/platform/telegraf/plugins/outputs
github.com/influxdata/platform
github.com/influxdata/platform/query/functions/inputs/storage
github.com/influxdata/platform/tsdb/cursors
github.com/influxdata/platform/pkg/snowflake
github.com/influxdata/platform/logger
github.com/influxdata/platform/pkg/slices
github.com/influxdata/platform/storage/reads/datatypes
github.com/influxdata/platform/pkg/binaryutil
github.com/influxdata/platform/pkg/bytesutil
github.com/influxdata/platform/pkg/mmap
github.com/influxdata/platform/pkg/rhh
github.com/influxdata/platform/query
github.com/influxdata/platform/tsdb
github.com/influxdata/platform/storage/reads
```

I still believe it would likely be easier to do the first solution.

To clarify, the issue arises because v1 uses code from v2 and v2 previously had a different import path which allowed us to have those be referenced simultaneously.

/cc @rbetts

**Comment by nathanielc** [comment]
Copying the code is not a solution, the code being shared is primarily the storage code, we want that code to be shared as it is being actively developed.

@jsternberg Can you clarify what would be disruptive about using `/v2` module path? Does that mean that all imports in the master branch need to use `/v2` in the path?

**Comment by stuartcarnie** [comment]
The current dependencies within `influxdb` are minimal, according to the following search, which only shows the 1.x branch depends on 5 distinct packages:

```sh
$ rg --glob '!Gopkg.*' -N --no-heading -F 'github.com/influxdata/platform' | gsort --key=2 | column -t
models/points.go:                        "github.com/influxdata/platform/models"
services/storage/series_cursor_test.go:  "github.com/influxdata/platform/models"
flux/functions/inputs/from.go:           "github.com/influxdata/platform/query/functions/inputs/storage"
services/storage/store.go:               "github.com/influxdata/platform/query/functions/inputs/storage"
flux/control/controller.go:              fstorage "github.com/influxdata/platform/query/functions/inputs/storage"
cmd/influxd/run/server.go:               "github.com/influxdata/platform/storage/reads"
internal/storage_store.go:               "github.com/influxdata/platform/storage/reads"
services/httpd/handler.go:               "github.com/influxdata/platform/storage/reads"
services/httpd/handler_test.go:          "github.com/influxdata/platform/storage/reads"
services/storage/predicate_test.go:      "github.com/influxdata/platform/storage/reads"
services/storage/series_cursor.go:       "github.com/influxdata/platform/storage/reads"
services/storage/store.go:               "github.com/influxdata/platform/storage/reads"
cmd/store/query/query.go:                "github.com/influxdata/platform/storage/reads/datatypes"
internal/storage_store.go:               "github.com/influxdata/platform/storage/reads/datatypes"
prometheus/converters.go:                "github.com/influxdata/platform/storage/reads/datatypes"
services/httpd/handler.go:               "github.com/influxdata/platform/storage/reads/datatypes"
services/httpd/handler_test.go:          "github.com/influxdata/platform/storage/reads/datatypes"
services/storage/predicate_test.go:      "github.com/influxdata/platform/storage/reads/datatypes"
services/storage/series_cursor.go:       "github.com/influxdata/platform/storage/reads/datatypes"
services/storage/store.go:               "github.com/influxdata/platform/storage/reads/datatypes"
tsdb/cursor.go:                          "github.com/influxdata/platform/tsdb/cursors"
```

My recommendation is to pull these packages into the 1.x branch to break the dependency permanently. 

There is no guarantee that the `storage/reads` or `tsdb/cursors` packages will maintain compatibility going forward, as we make improvements to these for the 2.x world, resulting in further complications.

**NOTE**: I have removed the dependency on `github.com/influxdata/platform/pkg/data/gen`. References to that package only exist in the Gopkg file.

**Comment by nathanielc** [comment]
@stuartcarnie Breaking the dependency for the `storage/reads` package means we have to double implement changes to that package. We have lots of planned work for the `storage/reads` package (i.e. the meta query bits). Is double implementing that work acceptable? Or would that work not ever make it to 1.x because it would require changes to tsm that are not being ported back?

**Comment by stuartcarnie** [comment]
@nathanielc I agree it is not ideal and I cringe at the thought I am even advocating such an approach. Your thinking is correct, we can't guarantee there won't be dependent changes to the `tsi`, `tsdb` or `tsm1` packages which exist as separate source trees in 1.x and 2.x. I know of some recent changes (or possibly new ones) such as

* TSM files are no longer compatible with the tags changes (\x00 and \xff for measurement and field keys)
* We will be making additional changes to use typed SeriesIDs, which only exist in 2.x.

I believe this will lead to a fallacy that upgrades should be straightforward and therefore break spectacularly or have very different performance characteristics. I anticipate the work would be substantial enough that it should be evaluated and planned as such.

**Comment by jsternberg** [comment]
As an aside, if we are eventually going to tag influxdb with v2 and allow people to use it as a library, go modules requires us to use the `/v2` suffix anyway so we may want to rip that bandaid off now. Otherwise, we'll get a `+incompatible` to the version.

We can separate that to a separate issue though.

**Comment by stuartcarnie** [comment]
Ideally we’d replicate the directory / package structure in 1.x so that it may be easier to compare and back port simple changes. This may require a little refactoring in 1.x.

**Comment by nathanielc** [comment]
@jsternberg OK it has been decided to go forward with the copy method as described by @stuartcarnie.

**Comment by jsternberg** [comment]
I've been attempting the copy method and I think I'm getting a larger surface area than I anticipated. The issue doesn't seem to be the `storage/reads`, but `query/stdlib/influxdata/influxdb`. All of the code that integrates the storage reader into flux seems like it doesn't contain any v2 specific things so copying it ends up getting a larger surface area than just the specific v1 reader bits and create duplication of code in particular with planner rules.

If this were just copying `storage/*` I think it's rather simple, but the addition of the stdlib code makes me more inclined to believe that we should do the v2 suffix and use the v2 interfaces. Then we can have v1 specific storage reading code, but not have to copy over the other stuff.

**Comment by jsternberg** [comment]
We're not sure what parts of the planner need to be copied over and which parts are unique so we're going to copy those over too and figure out what's in common later.

**Comment by jsternberg** [comment]
I have made progress. So far I have found that the `storage/reads` directory (and datatypes under it) port over pretty cleanly. I also moved `tsdb/clusters` over and things basically worked as-is.

I took out a lot of the implementations for the flux portions of things so those will likely have to be redone. I expected those would have to be redone though since the interfaces contained within that don't make sense for influxdb.
